metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,503 @@
+            }
+          )
+      else
+        None
+
+    val maybeChainedCodeAction = for {
+      document <- buffers.get(path)
+      termApply <- maybeTree
+      termApplyArg <- termApply.args.headOption
+      indentation = getIndentForPos(termApplyArg.pos, document)
+    } yield codeActionWithApply(
+      path,
+      termApply,
+      indentation
+    )
+    maybeChainedCodeAction.flatten.toSeq
+
+  }
+
+  private def constructCodeAction(
+      forElementsList: List[Enumerator],
+      yieldTerm: Term,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+
+    val indentedElems = forElementsList
+      .map(
+        _.syntax
+          .split(Array('\n'))
+          .map(line => s"$indentation   $line")
+          .mkString("\n")
+      )
+      .mkString("\n")
+
+    val yieldTermIndentedString = yieldTerm.syntax
+      .split(Array('\n'))
+      .map(line => s"$indentation   $line")
+      .mkString("\n")
+
+    val forYieldString =
+      s"""|{
+          |$indentation for {
+          |$indentedElems
+          |$indentation }  yield {
+          |$yieldTermIndentedString
+          |$indentation }
+          |$indentation}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def findTopMostApply(
+      currentTermApply: Term.Apply,
+      lastValidTermApply: Option[Term.Apply]
+  ): Option[Term.Apply] = {
+
+    currentTermApply.fun match {
+      case term if term.isNot[Term.Select] => lastValidTermApply
+      case termSelect: Term.Select
+          if termSelect.name.value != "map" && termSelect.name.value != "flatMap" &&
+            termSelect.name.value != "filter" && termSelect.name.value != "withFilter" && termSelect.name.value != "filterNot" =>
+        lastValidTermApply
+      case _: Term.Select =>
+        if (
+          !currentTermApply.parent.exists(parent =>
+            parent.is[Term.Select] && parent.parent.exists(_.is[Term.Apply])
+          )
+        ) {
+          Some(currentTermApply)
+        } else {
+          val result = Some(currentTermApply)
+          val grandParentApply =
+            currentTermApply.parent.get.parent.get.asInstanceOf[Term.Apply]
+          findTopMostApply(grandParentApply, result)
+        }
+    }
+  }
+
+  private def codeActionWithApply(
+      path: AbsolutePath,
+      termApply: Term.Apply,
+      indentation: String
+  ): Option[l.CodeAction] = {
+
+    findTopMostApply(termApply, None).flatMap { topMostApply =>
```

Comment:
What is `"topMost"Apply` ?

Reply:
this definitely needs some scaladoc. I think the `topMostApply` can be misleading. It is actually the `parentMostApply` or the `bottomMostApply` that is correct.

```
List(1, 2, 3)
  .m<<>>ap(x => x)
  .flatMap(Some(_))
  .filter(_ > 3)
```

Now when the cursor is on `map`, we want to start the conversion on `filter` instead, which is the `parentMost` or  `outerMost` apply.